### PR TITLE
Reimplement Masks + Fixes

### PR DESCRIPTION
**Pull request by IronApollo**
## Description
Masks Reimplemented:
    - Adjacent (~): Adjacent to a specific number of other blocks.
        - Example: ~[oak_log][1][4]
    - Extrema (#extrema): Restrict to near specific terrain extrema. The "-o" flag will only overlay existing terrain.
        - Example: #extrema[0d][45d][-o]
    - ROC Angle (#roc): Restrict to near specific terrain slope rate of change. The "-o" flag will only overlay existing terrain.
        - Example: #roc[0d][45d][-o]
    - Surface (#surface): Restrict to surfaces (any solid block touching air).
        - Example: #surface
    - Wall (#wall): Restrict to walls (any block n,e,s,w of air).
        - Example: #wall

Other Changes:
    - Minor logic change to angle mask's overlay.
    - Fixed negating air mask.
    - Added overlay flag to angle (/) mask.

### CODE REVIEW

File: worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/mask/AdjacentMaskParser.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+public class AdjacentMaskParser extends RichParser<Mask> {
+
+    public AdjacentMaskParser(WorldEdit worldEdit) {
+        super(worldEdit,"~");
+    }
+
+    @Override
+    protected Stream<String> getSuggestions(String argumentInput, int index) {
+        if(index == 0){
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/mask/AdjacentMaskParser.java:22]:
```java
if (index == 0) {
...
```
Plus a few other spots

File: worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/mask/RadiusMaskParser.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        }
+        return Stream.empty();
+    }
+
+    @Override
+    protected Mask parseFromInput(@NotNull String[] arguments, ParserContext context) throws InputParseException {
+        if(arguments.length < 2) return null;
+        int min = Integer.parseInt(arguments[0]);
+        int max = Integer.parseInt(arguments[1]);
+        System.out.println("radius: " + min + " | " + max);
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/mask/RadiusMaskParser.java:32]: Better shift this to a debug to keep it or remove it before merging.

File: worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/MaskFactory.java [hunk 3]
```diff
@@ -20,27 +20,7 @@
...
-import com.sk89q.worldedit.extension.factory.parser.mask.NoiseMaskParser;
-import com.sk89q.worldedit.extension.factory.parser.mask.OffsetMaskParser;
-import com.sk89q.worldedit.extension.factory.parser.mask.RegionMaskParser;
-import com.sk89q.worldedit.extension.factory.parser.mask.SimplexMaskParser;
-import com.sk89q.worldedit.extension.factory.parser.mask.SolidMaskParser;
-import com.sk89q.worldedit.extension.factory.parser.mask.TrueMaskParser;
-import com.sk89q.worldedit.extension.factory.parser.mask.XAxisMaskParser;
-import com.sk89q.worldedit.extension.factory.parser.mask.YAxisMaskParser;
-import com.sk89q.worldedit.extension.factory.parser.mask.ZAxisMaskParser;
+import com.sk89q.worldedit.extension.factory.parser.mask.*;
```
[the remark this answers is not in the document]
Comment (IronApollo) [reply worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/MaskFactory.java:23]: IDE does it automatically lol sorry

File: worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/mask/RadiusMaskParser.java (hunk 2 above)
Comment (IronApollo) [reply worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/mask/RadiusMaskParser.java:32]: Slipped through, will remove

File: worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/mask/AdjacentMaskParser.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        int min = arguments.length > 1 ? Integer.parseInt(arguments[1]) : -1;
+        int max = arguments.length > 2 ? Integer.parseInt(arguments[2]) : -1;
+        if (min == -1 && max == -1) {
+            min = 1;
+            max = 8;
+        } else if (max == -1) max = min;
+        if (max >= 8 && min == 1) {
+            return new AdjacentAnyMask(subMask);
+        }
+        return new AdjacentMask(subMask, (int) min, (int) max);
```
Comment (SirYwell) [inline worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/mask/AdjacentMaskParser.java:43]: The casts are redundant here

**Review by SirYwell — CHANGES_REQUESTED** [review]
Just minor stuff.

I found 2 flaws in my suggestion logic, I'll address them in a seperate small pr or you could fix it with that pr:
```patch
Index: worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/RichParser.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/RichParser.java	(revision 6807ccd960b3d761971ac74cb69d6f2081f04c2a)
+++ worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/RichParser.java	(date 1599508559065)
@@ -36,7 +36,7 @@
     @Override
     public Stream<String> getSuggestions(String input) {
         // we don't even want to start suggesting if it's not meant to be this parser result
-        if (input.length() > this.required.length() && !input.startsWith(this.required)) {
+        if (input.length() >= this.required.length() && !input.startsWith(this.required)) {
             return Stream.empty();
         }
         // suggest until the first [ as long as it isn't fully typed
@@ -45,11 +45,11 @@
         }
         // we know that it is at least "<required>"
         String[] strings = extractArguments(input.substring(this.prefix.length()), false);
-        StringJoiner joiner = new StringJoiner(",");
+        StringBuilder builder = new StringBuilder();
         for (int i = 0; i < strings.length - 1; i++) {
-            joiner.add("[" + strings[i] + "]");
+            builder.append('[').append(strings[i]).append(']');
         }
-        String previous = this.prefix + joiner;
+        String previous = this.prefix + builder;
         return getSuggestions(strings[strings.length - 1], strings.length - 1).map(s -> previous + "[" + s + "]");
     }
 

```

**Review by SirYwell — APPROVED** [review]

**Review by aurorasmiles — APPROVED** [review]

**Review by MattBDev — APPROVED** [review]

**Comment by TheoRetisch1** [comment]
What is the difference between the extrema and the roc mask? I have tested a little and can't find the difference?
Example: 
/gmask #extrema[45d][85d]
extrema

/gmask #roc[45d][85d]
roc

**Pull request merged**